File: modules/java/lib/main/src/main/java/com/mooltiverse/oss/nyx/data/Message.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.data;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * This object is a Git commit message value holder independent from the underlying Git implementation.
 */
public class Message {
    /**
     * The message footer lines, where keys are names and values are values.
     */
    private final Map<String,String> footers;

    /**
     * The full message.
     */
    private final String fullMessage;

    /**
     * The short message.
     */
    private final String shortMessage;

    /**
     * Constructor.
     * 
     * @param fullMessage the full message. Cannot be {@code null}
     * @param shortMessage the short message. Cannot be {@code null}
     * @param footers the map of message footers, where keys are names and values are values. Cannot be {@code null}
     */
    public Message(String fullMessage, String shortMessage, Map<String,String> footers) {
        super();
        Objects.requireNonNull(fullMessage);
        Objects.requireNonNull(shortMessage);
        Objects.requireNonNull(footers);
        this.fullMessage = fullMessage;
        this.shortMessage = shortMessage;
        this.footers = Collections.unmodifiableMap(footers);
    }

    /**
     * Returns the immutable list of footers, where keys are names and values are values.
     * 
     * @return the immutable list of footers, where keys are names and values are values. May be empty but not {@code null}.
     */
    public Map<String,String> getFooters() {
        return footers;
    }

    /**
     * Returns the full message.
     * 
     * @return the full message. Never {@code null}.
     */
    public String getFullMessage() {
        return fullMessage;
    }

    /**
     * Returns the short message.
     * 
     * @return the short message. Never {@code null}.
     */
    public String getShortMessage() {
        return shortMessage;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return shortMessage.concat(shortMessage.equals(fullMessage) ? "" : " ...");
    }
}

File: modules/go/nyx/entities/git/message.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package git

/*
This object is a Git commit message value holder independent from the underlying Git implementation.

This structure is JSON and YAML aware so all objects are properly managed for marshalling and unmarshalling. This comes with a downside
as all internal fields must be exported (have the first capital letter in their names) or they can't be marshalled.
*/
type Message struct {
	// The message footer lines, where keys are names and values are values.
	Footers map[string]string `json:"footers,omitempty" yaml:"footers,omitempty"`

	// The full message.
	FullMessage string `json:"fullMessage,omitempty" yaml:"fullMessage,omitempty"`

	// The short message.
	ShortMessage string `json:"shortMessage,omitempty" yaml:"shortMessage,omitempty"`
}

/*
Standard constructor.

Arguments are as follows:

- fullMessage the full message
- shortMessage the short message
- footers the map of message footers, where keys are names and values are values
*/
func NewMessageWith(fullMessage string, shortMessage string, footers map[string]string) *Message {
	m := Message{}

	m.FullMessage = fullMessage
	m.ShortMessage = shortMessage
	m.Footers = footers

	return &m
}

/*
Returns the immutable list of footers, where keys are names and values are values.
*/
func (m Message) GeFooters() map[string]string {
	return m.Footers
}

/*
Returns the full message.
*/
func (m Message) GetFullMessage() string {
	return m.FullMessage
}

/*
Returns the short message.
*/
func (m Message) GetShortMessage() string {
	return m.ShortMessage
}

/*
Returns the string representation of the message.
*/
func (m Message) String() string {
	if m.FullMessage == "" {
		return m.ShortMessage
	} else {
		return m.ShortMessage + " ..."
	}
}
